toolchain/lower/file_context.h

```diff
@@ -157,6 +188,50 @@ class FileContext {
 
   auto BuildVtable(const SemIR::Class& class_info) -> llvm::GlobalVariable*;
 
+  // Function is used to track which specifics were lowered for each generic.
+  // These are added one by one while lowering their definitions.
+  auto AddLoweredSpecificForGeneric(SemIR::GenericId generic_id,
+                                    SemIR::SpecificId specific_id) {
+    lowered_specifics_[generic_id.index].push_back(specific_id);
+  }
+  // Initialize a return a SpecificFunctionFingerprint* instance for specific.
+  // The internal of the fingerprint are populated during and after lowering
+  // the function body of that specific.
+  auto InitializeFingerprintForSpecific(SemIR::SpecificId specific_id)
+      -> SpecificFunctionFingerprint* {
+    if (!specific_id.has_value()) {
+      return nullptr;
+    }
+    return &lowered_specific_fingerprint_[specific_id.index];
+  }
+
+  // Entry point for coalescing equivalent specifics.
+  auto CoalesceEquivalentSpecifics() -> void;
+  // While coalescing specifics, compare the function types for two specifics.
+  // This uses a fingerprint generated for each function type.
+  auto AreFunctionTypesEquivalent(SemIR::SpecificId specific1,
+                                  SemIR::SpecificId specific2) -> bool;
```

Comment:
WDYT of something like `lhs_id` and `rhs_id`, or something like that? Kind of getting at two things:

1. A lot of the toolchain code suffixes "_id" when we have an id, partly to disambiguate from the type referenced by an ID.
2. "specific_id1" would be fine, though I have a slight leaning towards non-indexed names when it's not N-ary, and lhs/rhs is a common way of doing that.

Reply:
I left all with the format `specific_id1` for consistency. I'd prefer something shorter (and not a fan of using the same name in a different scope, which happens now). The `lhs_id` doesn't really fit, there's not left and right, just a pair of two ids being compared, more akin to "first"/"second" in a pair. I'd like to do a renaming of all (again, consistency) given a better name, but I don't really have a compelling one. I'm flexible if you have a stronger preference for another naming.